src/treq/test/httpbin_server.py

```diff
@@ -0,0 +1,224 @@
+"""
+A local ``httpbin`` server to run integration tests against.
+
+This ensures tests do not depend on `httpbin <https://httpbin.org/>`_.
+"""
+from __future__ import print_function
+import attr
+import argparse
+import datetime
+import json
+import sys
+
+import httpbin
+
+from twisted.internet.defer import Deferred
+from twisted.internet.endpoints import TCP4ServerEndpoint, SSL4ServerEndpoint
+from twisted.internet.task import react
+from twisted.internet.ssl import (Certificate,
+                                  CertificateOptions)
+
+from OpenSSL.crypto import PKey, X509
+
+from twisted.python.threadpool import ThreadPool
+from twisted.web.server import Site
+from twisted.web.wsgi import WSGIResource
+
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+from cryptography.x509.oid import NameOID
+from cryptography.hazmat.primitives.serialization import Encoding
+
+
+@attr.s
+class _HTTPBinDescription(object):
+    """
+    Describe an ``httpbin`` process.
+
+    :param host: The host on which the process listens.
+    :type host: :py:class:`str`
+
+    :param port: The port on which the process listens.
+    :type port: :py:class:`int`
+
+    :param cacert: (optional) The PEM-encoded certificate authority's
+        certificate.  The calling process' treq must trust this when
+        running HTTPS tests.
+    :type cacert: :py:class:`str` or :py:class:`None`
+    """
+    host = attr.ib()
+    port = attr.ib()
+    cacert = attr.ib(default=None)
+
+    @classmethod
+    def from_dictionary(cls, dictionary):
+        return cls(**dictionary)
+
+
+def certificatesForAuthorityAndServer(serviceIdentity):
+    """
+    Create a self-signed CA certificate and server certificate signed
+    by the CA.
+
+    @param serviceIdentity: The identity (hostname) of the server.
+    @type serviceIdentity: L{unicode}
+
+    @return: a 3-tuple of C{(certificate_authority_certificate,
+        server_private_key, server_certificate)}
+    @rtype: L{tuple} of (L{sslverify.Certificate},
+        L{OpenSSL.crypto.PKey, OpenSSL.crypto.X509})
+    """
+    commonNameForCA = x509.Name(
+        [x509.NameAttribute(NameOID.COMMON_NAME, u'Testing Example CA')]
+    )
+    commonNameForServer = x509.Name(
+        [x509.NameAttribute(NameOID.COMMON_NAME, u'Testing Example Server')]
+    )
+    oneDay = datetime.timedelta(1, 0, 0)
+    privateKeyForCA = rsa.generate_private_key(
+        public_exponent=65537,
+        key_size=1024,
```

Comment:
This justification should really be hoisted out and documented (i.e.: `MINIMUM_KEY_SIZE` or something with a `@var` containing some text similar to your comment, not `1024`)